compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp

```diff
@@ -0,0 +1,164 @@
+  moduleSymbols.insert(newGlobal);
+  SymbolTable::setSymbolVisibility(newGlobal,
+                                   SymbolTable::getSymbolVisibility(global));
+
+  // Rewrite loads and stores to use the new global.
+  auto expandShapeReInds =
+      getReassociationIndicesForReshape(globalType, newGlobalType);
+  if (!expandShapeReInds) {
+    return failure();
+  }
+  for (auto loadOrStore : loadStoreOps) {
+    rewriter.setInsertionPoint(loadOrStore);
+    if (auto load = dyn_cast<IREE::Util::GlobalLoadOpInterface>(loadOrStore)) {
+      auto newOp = rewriter.create(
+          load->getLoc(), load->getName().getIdentifier(), load->getOperands(),
+          {newGlobalType}, load->getAttrs());
+      auto newLoad = dyn_cast<IREE::Util::GlobalLoadOpInterface>(newOp);
+      newLoad.setGlobalAttr(FlatSymbolRefAttr::get(newGlobal.getGlobalName()));
+      newLoad.setGlobalImmutable(load.isGlobalImmutable());
+      rewriter.replaceOpWithNewOp<tensor::ExpandShapeOp>(
+          load, globalType, newLoad->getResult(0), expandShapeReInds.value());
+    } else if (auto store =
+                   dyn_cast<IREE::Util::GlobalStoreOpInterface>(loadOrStore)) {
+      Value collapse = rewriter.create<tensor::CollapseShapeOp>(
+          store.getLoc(), newGlobalType, store->getOperand(0),
+          expandShapeReInds.value());
+      auto newOp = rewriter.create(
+          store->getLoc(), store->getName().getIdentifier(),
+          store->getOperands(), store->getResultTypes(), store->getAttrs());
+      auto newStore = dyn_cast<IREE::Util::GlobalStoreOpInterface>(newOp);
+      newStore.setGlobalAttr(FlatSymbolRefAttr::get(newGlobal.getGlobalName()));
+      newStore.setStoredGlobalValue(collapse);
+      rewriter.eraseOp(store);
+    } else {
+      return failure();
+    }
+  }
+  rewriter.eraseOp(global);
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Pass definition
+//===----------------------------------------------------------------------===//
+
+class FoldGlobalUnitDimsPass
+    : public FoldGlobalUnitDimsBase<FoldGlobalUnitDimsPass> {
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    DenseMap<StringRef, SmallVector<Operation *>> loadStoreMap;
+    auto addToLoadStoreMap = [&](StringRef name, Operation *loadStoreOp) {
+      if (loadStoreMap.contains(name)) {
+        loadStoreMap[name].push_back(loadStoreOp);
+      } else {
+        SmallVector<Operation *> loadStores(1, loadStoreOp);
+        loadStoreMap.insert(std::make_pair(name, loadStores));
+      }
+    };
+    moduleOp.walk([&](Operation *op) {
+      if (auto load = dyn_cast<IREE::Util::GlobalLoadOpInterface>(op)) {
+        addToLoadStoreMap(load.getGlobalName(), op);
+      } else if (auto store =
+                     dyn_cast<IREE::Util::GlobalStoreOpInterface>(op)) {
+        addToLoadStoreMap(store.getGlobalName(), op);
+      }
+    });
```

Comment:
Can you use the Explorer to set up this map? It might be overkill in this case, and you don't need the inverse call graph, but I'd guess that the code would look more canonical? You can use `Explorer::initializeGlobalInfos` on its own to initialize the global information without setting up the callgraph I think.

Also provides nice APIs like `forEachGlobal` and caches information like mutability, indirect accessors, and visibility.